File: I, Robot Emulator 1998/Mathbox.c
// Copyright 1997, 1998, 2020 by John Manfreda. All Rights Reserved.
// https://www.manfreda.org/
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.If not, see<https://www.gnu.org/licenses/>.

typedef unsigned short WORD;
typedef unsigned char BYTE;

// external variable needed
extern WORD * MathRAMROM;

// external function prototypes
void GetCamera( void );
void DoObject( int ObjAddr );
void RenderPlayfield( void );

#define StartPlayfield 0x8400
#define UnknownMathFunction 0x8600
#define Roll 0x8800
#define Yaw 0x9000
#define Pitch 0xA000
#define Camera 0xC000

#pragma check_stack(off)

// Pitch function  (X)
// Multiplies source matrix by
//	[ 1   0   0  ]
//	[ 0  cos sin ]
//	[ 0 -sin cos ]
static void _inline Pfunction( short * matrix, short SIN, short COS )
{
	short B, C, E, F, H, I;

	B = matrix[1];
	C = matrix[2];
	E = matrix[4];
	F = matrix[5];
	H = matrix[7];
	I = matrix[8];

	matrix[1] = (short)((B * COS - C * SIN) >> 14);
	matrix[2] = (short)((C * COS + B * SIN) >> 14);
	matrix[4] = (short)((E * COS - F * SIN) >> 14);
	matrix[5] = (short)((F * COS + E * SIN) >> 14);
	matrix[7] = (short)((H * COS - I * SIN) >> 14);
	matrix[8] = (short)((I * COS + H * SIN) >> 14);
}

// Yaw function  (Y)
// Multiplies source matrix by
//	[ cos 0 -sin ]
//	[  0  1   0  ]
//	[ sin 0  cos ]
static void _inline Yfunction( short * matrix, short SIN, short COS )
{
	short A, C, D, F, G, I;

	A = matrix[0];
	C = matrix[2];
	D = matrix[3];
	F = matrix[5];
	G = matrix[6];
	I = matrix[8];
	
	matrix[0] = (short)((A * COS + C * SIN) >> 14);
	matrix[2] = (short)((C * COS - A * SIN) >> 14);
	matrix[3] = (short)((D * COS + F * SIN) >> 14);
	matrix[5] = (short)((F * COS - D * SIN) >> 14);
	matrix[6] = (short)((G * COS + I * SIN) >> 14);
	matrix[8] = (short)((I * COS - G * SIN) >> 14);
}

// Roll function  (Z)
// Multiplies source matrix by
//	[  cos sin 0 ]
//	[ -sin cos 0 ]
//	[   0   0  1 ]
static void _inline Rfunction( short * matrix, short SIN, short COS )
{
	short A, B, D, E, G, H;

	A = matrix[0];
	B = matrix[1];
	D = matrix[3];
	E = matrix[4];
	G = matrix[6];
	H = matrix[7];
	
	matrix[0] = (short)((A * COS - B * SIN) >> 14);
	matrix[1] = (short)((B * COS + A * SIN) >> 14);
	matrix[3] = (short)((D * COS - E * SIN) >> 14);
	matrix[4] = (short)((E * COS + D * SIN) >> 14);
	matrix[6] = (short)((G * COS - H * SIN) >> 14);
	matrix[7] = (short)((H * COS + G * SIN) >> 14);
}


// Camera function
// performs A x B = C
// result is C transpose
static void _inline Cfunction( short * matrixA, short * matrixB, short * matrixC )
{
	matrixC[0] = (short)((matrixA[0]*matrixB[0] + matrixA[1]*matrixB[1] + matrixA[2]*matrixB[2]) >> 14);
	matrixC[1] = (short)((matrixA[0]*matrixB[3] + matrixA[1]*matrixB[4] + matrixA[2]*matrixB[5]) >> 14);
	matrixC[2] = (short)((matrixA[0]*matrixB[6] + matrixA[1]*matrixB[7] + matrixA[2]*matrixB[8]) >> 14);
	matrixC[3] = (short)((matrixA[3]*matrixB[0] + matrixA[4]*matrixB[1] + matrixA[5]*matrixB[2]) >> 14);
	matrixC[4] = (short)((matrixA[3]*matrixB[3] + matrixA[4]*matrixB[4] + matrixA[5]*matrixB[5]) >> 14);
	matrixC[5] = (short)((matrixA[3]*matrixB[6] + matrixA[4]*matrixB[7] + matrixA[5]*matrixB[8]) >> 14);
	matrixC[6] = (short)((matrixA[6]*matrixB[0] + matrixA[7]*matrixB[1] + matrixA[8]*matrixB[2]) >> 14);
	matrixC[7] = (short)((matrixA[6]*matrixB[3] + matrixA[7]*matrixB[4] + matrixA[8]*matrixB[5]) >> 14);
	matrixC[8] = (short)((matrixA[6]*matrixB[6] + matrixA[7]*matrixB[7] + matrixA[8]*matrixB[8]) >> 14);
}

void DoMathbox( void )
{
	WORD command = *MathRAMROM;

	switch( command )
	{
	case StartPlayfield:
		GetCamera();
		RenderPlayfield( );
		break;
	case UnknownMathFunction:
		break;
	case Roll:
		Rfunction( &MathRAMROM[MathRAMROM[6]], MathRAMROM[7], MathRAMROM[8] );
		break;
	case Yaw:
		Yfunction( &MathRAMROM[MathRAMROM[6]], MathRAMROM[7], MathRAMROM[8] );
		break;
	case Pitch:
		Pfunction( &MathRAMROM[MathRAMROM[6]], MathRAMROM[7], MathRAMROM[8] );
		break;
	case Camera:
		Cfunction( &MathRAMROM[MathRAMROM[0xF6/2]], &MathRAMROM[MathRAMROM[0xF8/2]], &MathRAMROM[MathRAMROM[0xFA/2]] );
		break;
	default:
		GetCamera();
		DoObject( command );
		break;
	}
}

File: I, Robot Emulator MonoGame/Hardware/Subsystems/Mathbox.cs
// Copyright 2020 by John Manfreda. All Rights Reserved.
// https://www.manfreda.org/
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.If not, see<https://www.gnu.org/licenses/>.

using System;
using System.Runtime.InteropServices;
using System.Runtime.Serialization;

namespace I_Robot
{
    /// <summary>
    /// Represents the Mathbox hardware on the I, Robot PCB
    /// 
    /// The Mathbox is capable of performing the following operations
    /// - Yaw a matrix by an angle
    /// - Pitch a matrix by an angle
    /// - Roll a matrix by an angle
    /// - Perform a fast matrix multiply
    /// - converts playfield data into a set of display list commands for the video processor
    /// </summary>
    [Serializable]
    unsafe public class Mathbox : Hardware.Subsystem
    {
        public interface IInterpreter
        {
            /// <summary>
            /// Sets pointer to mathbox memory
            /// </summary>
            public UInt16 * pMemory { set; }

            /// <summary>
            /// Sets the current video buffer being renderd to
            /// </summary>
            /// <param name="index"></param>
            public void SetVideoBuffer(int index);

            /// <summary>
            /// Erases the selected video buffer
            /// </summary>
            public void EraseVideoBuffer();

            /// <summary>
            /// This command kicks off playfield rasterization
            /// </summary>
            void RasterizePlayfield();

            /// <summary>
            /// This command performas an unknown function, likely related to rasterization
            /// </summary>
            void UnknownCommand();

            /// <summary>
            /// Rasterizes the object at the given address, using the current lighting vector, camera vector, etc
            /// </summary>
            /// <param name="address"></param>
            void RasterizeObject(UInt16 address);
        }

        /// <summary>
        /// enumeration of commands that are recognized by the Mathbox
        /// </summary>
        public enum COMMAND : UInt16
        {
            START_PLAYFIELD = 0x8400,
            UNKNOWN = 0x8600,
            ROLL = 0x8800,
            YAW = 0x9000,
            PITCH = 0xA000,
            MATRIX_MULTIPLY = 0xC000,
        }

        /// <summary>
        /// Represents a 3x3 matrix as used by the Mathbox
        /// </summary>
        [StructLayout(LayoutKind.Sequential, Pack = 2)]
        public struct Matrix
        {
            public Int16 M11, M12, M13;
            public Int16 M21, M22, M23;
            public Int16 M31, M32, M33;

            public override string ToString()
            {
                return $"| {M11.HexString()}, {M12.HexString()}, {M13.HexString()} |\n| {M21.HexString()}, {M22.HexString()}, {M23.HexString()} |\n| {M31.HexString()}, {M32.HexString()}, {M33.HexString()} |";
            }
        }

        /// <summary>
        /// Patches self test to allow it to pass mathbox testing
        /// Replaces the result matrices with results that this mathbox generates
        /// </summary>
        static public readonly byte[] SelfTestPatch = new byte[]
               { 0x40,0x00, 0x00,0x00, 0x01,0x21, // yaw test result
                  0x00,0x00, 0x40,0x00, 0x00,0x00,
                  0xFE,0xE1, 0x00,0x00, 0x40,0x01,
                  0x40,0x01, 0xFE,0xE1, 0x01,0x21, // roll test result
                   0x01,0x21, 0x40,0x00, 0x00,0x00,
                  0xFE,0xE9, 0x00,0x05, 0x40,0x01,
                  0x40,0x01, 0xFE,0xE8, 0x01,0x2C, // pitch test result
                  0x01,0x21, 0x40,0x01, 0xFE,0xE1,
                  0xFE,0xE9, 0x01,0x22, 0x3F,0xFC,
                  0x40,0x01, 0x01,0x21, 0xFE,0xE9, // matrix multiply result
                  0xFE,0xE8, 0x40,0x01, 0x01,0x22,
                  0x01,0x2C, 0xFE,0xE1, 0x3F,0xFC };

        // 6 mathbox ROM banks that are accessible by the CPU
        public readonly PinnedBuffer<byte>[] ROM = new PinnedBuffer<byte>[6] {
            new PinnedBuffer<byte>(0x2000),
            new PinnedBuffer<byte>(0x2000),
            new PinnedBuffer<byte>(0x2000),
            new PinnedBuffer<byte>(0x2000),
            new PinnedBuffer<byte>(0x2000),
            new PinnedBuffer<byte>(0x2000) };

        /// <summary>
        /// A full 32k x 16 bits of Mathbox memory
        /// Lower 8k x 16 bits of memory are RAM
        /// remaining memory is from mathbox ROMs
        /// </summary>
        readonly PinnedBuffer<UInt16> Memory16 = new PinnedBuffer<UInt16>(0x8000);

        /// <summary>
        /// Byte pointer to mathbox memory
        /// </summary>
        readonly byte* pData;

        /// <summary>
        /// Word pointer to mathbox memory
        /// </summary>
        readonly UInt16* Memory;

        /// <summary>
        /// Interface to provided interpreter that intercepts rendering commands
        /// </summary>
        readonly IInterpreter Interpreter;

        bool mMATH_START = false;
        bool mERASE = false;
        bool ?mBUFSEL = false;

        /// <summary>
        /// Hardware signal that indicates when the mathbox is done
        /// </summary>
        public bool MB_DONE { get; private set; } = true;

        // keep delegates around to prevent them from being garbage collected
        public readonly M6809E.ReadDelegate ReadRamFunction;
        public readonly M6809E.WriteDelegate WriteRamFunction;

        public Mathbox(Hardware hardware, IInterpreter interpreter) : base(hardware, "Mathbox")
        {
            Memory = Memory16;
            pData = (byte*)Memory;

            Interpreter = interpreter;
            Interpreter.pMemory = Memory;

            //  Mathbox     CPU           hi                lo
            //  2000-2FFF   0:2000-3FFF   136029-104[0000]  136029-103[0000]
            //  3000-3FFF   1:2000-3FFF   136029-104[1000]  136029-103[1000]
            //  4000-4FFF   2:2000-3FFF   136029-102[0000]  136029-101[0000]
            //  5000-5FFF   3:2000-3FFF   136029-102[1000]  136029-101[1000]
            //  6000-6FFF   4:2000-3FFF   136029-102[2000]  136029-101[2000]
            //  7000-7FFF   5:2000-3FFF   136029-102[3000]  136029-101[3000]

            ROM? r101 = hardware.Roms["136029-101"];
            ROM? r102 = hardware.Roms["136029-102"];
            ROM? r103 = hardware.Roms["136029-103"];
            ROM? r104 = hardware.Roms["136029-104"];

            // create x86 accessable bank (low endian)
            UInt16 address;
            if (r103 != null)
            {
                address = 0x2000;
                foreach (byte b in r103)
                    Memory[address++] = b;
            }
            if (r104 != null)
            {
                address = 0x2000;
                foreach (byte b in r104)
                    Memory[address++] |= (UInt16)(b << 8);
            }
            if (r101 != null)
            {
                address = 0x4000;
                foreach (byte b in r101)
                    Memory[address++] = b;
            }
            if (r102 != null)
            {
                address = 0x4000;
                foreach (byte b in r102)
                    Memory[address++] |= (UInt16)(b << 8);
            }
            
            // create M6809E accessable banks (high endian)
            address = 0x2000;
            for (int bank = 0; bank < ROM.Length; bank++)
            {
                for (int n = 0; n < 0x2000;)
                {
                    ROM[bank][n++] = (byte)(Memory[address] >> 8);
                    ROM[bank][n++] = (byte)(Memory[address++] >> 0);
                }
            }

#if true
            // now patch the self test so our emulation results passs
            Array.Copy(Mathbox.SelfTestPatch, 0, Hardware.ProgROM.Bank_4000[4], 0x5CAD - 0x4000, SelfTestPatch.Length);

            // adjust the value at the end of ROM bank 4 to account for the checksum change
            UInt16 checksum = 0;
            for (int n = 0; n < 0x1FFE; n += 2)
                checksum -= (UInt16)(Hardware.ProgROM.Bank_4000[4][n] * 256 + Hardware.ProgROM.Bank_4000[4][n + 1]);
            for (int n = 0; n < 0x2000; n += 2)
                checksum -= (UInt16)(Hardware.ProgROM.Bank_4000[5][n] * 256 + Hardware.ProgROM.Bank_4000[5][n + 1]);
            Hardware.ProgROM.Bank_4000[4][0x1FFE] = (byte)(checksum >> 8);
            Hardware.ProgROM.Bank_4000[4][0x1FFF] = (byte)(checksum >> 0);
#endif

            ReadRamFunction = new M6809E.ReadDelegate((UInt16 address) =>
            {
                // flip 6809 hi/lo, so x86 can read words
                return (byte)pData[(address & 0x1FFF) ^ 1];
            });

            WriteRamFunction = new M6809E.WriteDelegate((UInt16 address, byte data) =>
            {
                // flip 6809 hi/lo, so x86 can read words
                pData[(address & 0x1FFF) ^ 1] = data;
            });
        }

        public override void Dispose()
        {
        }

        public override void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            info.AddValue("MATH_START", MATH_START);
            info.AddValue("MB_DONE", MB_DONE);
            info.AddValue("RAM", Memory16);
        }

        public override void Reset()
        {
            MB_DONE = true;
            
            // NOTE: 6809 I/O is not handled here, it is handled by Bank_2000 subsystem
        }

        public bool BUFSEL
        {
            set
            {
                if (mBUFSEL != value)
                {
                    mBUFSEL = value;
                    Interpreter.SetVideoBuffer(value ? 1 : 0);
                }
            }
        }

        /// <summary>
        /// Hardware signal that tells the video process to erase the selected video bank
        /// </summary>
        public bool ERASE
        {
            set
            {
                if (mERASE != value)
                {
                    mERASE = value;
                    if (value)
                        Interpreter.EraseVideoBuffer();
                }
            }
        }

        /// <summary>
        /// Hardware signal that kicks off mathbox execution
        /// </summary>
        public bool MATH_START
        {
            get => mMATH_START;
            set
            {
                // if signal is changing
                if (mMATH_START != value)
                {
                    mMATH_START = value;

                    // if signal is going from 0 --> 1
                    if (value)
                    {
                        // start execution
                        // command is in first mathbox address
                        switch ((COMMAND)Memory[0])
                        {
                            case COMMAND.START_PLAYFIELD:
                                // route to interpreter
                                Interpreter.RasterizePlayfield();
                                break;
                            case COMMAND.UNKNOWN:
                                // route to interpreter
                                Interpreter.UnknownCommand();
                                break;
                            case COMMAND.ROLL:
                                // native interpretation
                                Roll((Matrix*)&Memory[Memory[6]], (Int16)Memory[7], (Int16)Memory[8]);
                                break;
                            case COMMAND.YAW:
                                // native interpretation
                                Yaw((Matrix*)&Memory[Memory[6]], (Int16)Memory[7], (Int16)Memory[8]);
                                break;
                            case COMMAND.PITCH:
                                // native interpretation
                                Pitch((Matrix*)&Memory[Memory[6]], (Int16)Memory[7], (Int16)Memory[8]);
                                break;
                            case COMMAND.MATRIX_MULTIPLY:
                                // native interpretation
                                MatrixMultiply((Matrix*)&Memory[Memory[0x7B]], (Matrix*)&Memory[Memory[0x7C]], (Matrix*)&Memory[Memory[0x7D]]);
                                break;
                            default:
                                // route to interpreter
                                Interpreter.RasterizeObject(Memory[0]);
                                break;
                        }

                        // signal mathbox completion to CPU
                        Hardware.M6809E.FIRQ = true;
                    }
                }
            }
        }

        /// <summary>
        /// Pitch function (rotate about X axis).
        ///
        ///                     [ 1   0   0  ]
        ///  matrix = matrix x	[ 0  cos sin ]
        ///                     [ 0 -sin cos ]
        /// </summary>
        /// <param name="pMatrix">pointer to matrix</param>
        /// <param name="sin">sin() of angle to rotate by</param>
        /// <param name="cos">cos() of angle to rotate by</param>
        void Pitch(Matrix* pMatrix, Int16 sin, Int16 cos)
        {
            Int16 m12 = pMatrix->M12, m13 = pMatrix->M13;
            Int16 m22 = pMatrix->M22, m23 = pMatrix->M23;
            Int16 m32 = pMatrix->M32, m33 = pMatrix->M33;

            pMatrix->M12 = (Int16)((m12 * cos - m13 * sin) >> 14);
            pMatrix->M13 = (Int16)((m13 * cos + m12 * sin) >> 14);
            pMatrix->M22 = (Int16)((m22 * cos - m23 * sin) >> 14);
            pMatrix->M23 = (Int16)((m23 * cos + m22 * sin) >> 14);
            pMatrix->M32 = (Int16)((m32 * cos - m33 * sin) >> 14);
            pMatrix->M33 = (Int16)((m33 * cos + m32 * sin) >> 14);
        }

        /// <summary>
        /// Yaw function (rotate about Y axis).
        ///
        ///                     [ cos 0 -sin ]
        ///  matrix = matrix x	[  0  1   0  ]
        ///                     [ sin 0  cos ]
        /// </summary>
        /// <param name="pMatrix">pointer to matrix</param>
        /// <param name="sin">sin() of angle to rotate by</param>
        /// <param name="cos">cos() of angle to rotate by</param>
        void Yaw(Matrix* pMatrix, Int16 sin, Int16 cos)
        {
            Int16 m11 = pMatrix->M11, m13 = pMatrix->M13;
            Int16 m21 = pMatrix->M21, m23 = pMatrix->M23;
            Int16 m31 = pMatrix->M31, m33 = pMatrix->M33;

            pMatrix->M11 = (Int16)((m11 * cos + m13 * sin) >> 14);
            pMatrix->M13 = (Int16)((m13 * cos - m11 * sin) >> 14);
            pMatrix->M21 = (Int16)((m21 * cos + m23 * sin) >> 14);
            pMatrix->M23 = (Int16)((m23 * cos - m21 * sin) >> 14);
            pMatrix->M31 = (Int16)((m31 * cos + m33 * sin) >> 14);
            pMatrix->M33 = (Int16)((m33 * cos - m31 * sin) >> 14);
        }

        /// <summary>
        /// Roll function (rotate about Z axis)
        /// 
        ///                   [  cos sin 0 ]
        /// matrix = matrix x [ -sin cos 0 ]
        ///                   [   0   0  1 ]
        /// </summary>
        /// <param name="pMatrix">pointer to matrix</param>
        /// <param name="sin">sin() of angle to rotate by</param>
        /// <param name="cos">cos() of angle to rotate by</param>
        void Roll(Matrix* pMatrix, Int16 sin, Int16 cos)
        {
            Int16 m11 = pMatrix->M11, m12 = pMatrix->M12;
            Int16 m21 = pMatrix->M21, m22 = pMatrix->M22;
            Int16 m31 = pMatrix->M31, m32 = pMatrix->M32;

            pMatrix->M11 = (Int16)((m11 * cos - m12 * sin) >> 14);
            pMatrix->M12 = (Int16)((m12 * cos + m11 * sin) >> 14);
            pMatrix->M21 = (Int16)((m21 * cos - m22 * sin) >> 14);
            pMatrix->M22 = (Int16)((m22 * cos + m21 * sin) >> 14);
            pMatrix->M31 = (Int16)((m31 * cos - m32 * sin) >> 14);
            pMatrix->M32 = (Int16)((m32 * cos + m31 * sin) >> 14);
        }

        /// <summary>
        /// Matrix multiply 
        /// performs A x B = C
        /// result is C transpose
        /// </summary>
        /// <param name="pA">pointer to first matrix</param>
        /// <param name="pB">pointer to second matrix</param>
        /// <param name="pC">pointer to result matrix</param>
        void MatrixMultiply(Matrix* pA, Matrix* pB, Matrix* pC)
        {
            pC->M11 = (Int16)((pA->M11 * pB->M11 + pA->M12 * pB->M12 + pA->M13 * pB->M13) >> 14);
            pC->M12 = (Int16)((pA->M11 * pB->M21 + pA->M12 * pB->M22 + pA->M13 * pB->M23) >> 14);
            pC->M13 = (Int16)((pA->M11 * pB->M31 + pA->M12 * pB->M32 + pA->M13 * pB->M33) >> 14);
            pC->M21 = (Int16)((pA->M21 * pB->M11 + pA->M22 * pB->M12 + pA->M23 * pB->M13) >> 14);
            pC->M22 = (Int16)((pA->M21 * pB->M21 + pA->M22 * pB->M22 + pA->M23 * pB->M23) >> 14);
            pC->M23 = (Int16)((pA->M21 * pB->M31 + pA->M22 * pB->M32 + pA->M23 * pB->M33) >> 14);
            pC->M31 = (Int16)((pA->M31 * pB->M11 + pA->M32 * pB->M12 + pA->M33 * pB->M13) >> 14);
            pC->M32 = (Int16)((pA->M31 * pB->M21 + pA->M32 * pB->M22 + pA->M33 * pB->M23) >> 14);
            pC->M33 = (Int16)((pA->M31 * pB->M31 + pA->M32 * pB->M32 + pA->M33 * pB->M33) >> 14);
        }
    }
}